konveyor/pvc-migrate
# Improvements in Stage 2 around volume sizes on the destination cluster

Fixes #92
Fixes #111 

**This PR aims at solving two problems :** 
- Volumes that are grown outside of kube are not the same size as that of their associated PVCs. In current implementation of Stage 2, this case is not handled.
- Volumes that are almost full may end up requiring more space on the destination cluster due to changes in storage class. Current stage 2 implementation doesn't handle this case, too.

**New approach proposes :**
- On the destination side, we request the PVC with capacity equal to the maximum of following sizes:
  * actual used size + 5% (configureable value)
  * PVC requested size 
  * actual provisioned volume size

**Sizing conversion :**
PVC sizes use SI units `K,Ki,M,Mi,G,Gi,T,Ti` and so on. On the other hand, linux disk usage utilities have different units than PVCs. Also, the provisioned PV size in Kube is not exactly equal to that of reported by `df`. Our approach, thus, uses approximations around sizes.

Currently, supported size units : `M,Mi,G,Gi,T,Ti`  

Any PVC that does not have size mentioned in above units, will be skipped from the size adjustment phase.

**Performance considerations:** 

With the above approach, Stage 2 now needs to talk to the source cluster for every PV to make sure the sizes are correct. In current implementation, PVs are processed serially one-by-one on a per ns basis. That means, for `n` PVCs, we would have `n` SSH logins to different nodes. 

To optimize, we group all PVCs by node names, and process all PVCs on a node in one go, reducing the number of logins from `n` to the number of nodes. The output of Stage 1 is modified in-flight to include data about decisions that were taken to calculate the new size of the PVCs.

**How to test?**
[_Scanario 1_] PV almost full 

- Deploy pvc-migrate [Data Generator benchmark app](https://github.com/konveyor/pvc-migrate/tree/master/benchmarks/DataGenerator) by modifying the size of the PVC and the file size : 

```sh
ansible-playbook playbook.yml -e namespace=bmark-1 -e pvc_size=1Gi -e file_size=1000
```

[_Scenario 2_] Volume size & PVC size mismatch

- Deploy pvc-migrate [Data Generator benchmark app](https://github.com/konveyor/pvc-migrate/tree/master/benchmarks/DataGenerator) by modifying the size of the PVC and the file size :

```sh
ansible-playbook playbook.yml -e namespace=bmark-2 -e pvc_size=1100Mi -e file_size=100
```

Once deployed, run `Stage 1` and `Stage 2` for `bmark-1` and `bmark-2` namespaces. Verify the size of the PVCs on the destination.
